### PR TITLE
RGW - Zipper - remove RGWObjectCtx from SAL API

### PR DESCRIPTION
**Pull request by dang**
RGWObjectCtx is a RADOS-specific concept, but it's threaded through the
whole of RGW.  Remove it from the SAL API, and from everything above the
line, leaving it only in RadosStore.  As part of this, move RGWObjState
to SAL and use it as the key for an Object.

Signed-off-by: Daniel Gryniewicz <dang@fprintf.net>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4189,6 +4186,7 @@ void RGWPutObj::execute(optional_yield y)
     emplace_attr(RGW_ATTR_OBJECT_RETENTION, std::move(obj_retention_bl));
   }
 
+  ldpp_dout(this, 0) << "Luke 1 if_match=" << if_match << dendl;
```
Comment (soumyakoduri) [inline src/rgw/rgw_op.cc:4189]: nit: cleanup

File: src/rgw/store/dbstore/common/dbstore.cc [hunk 2]
```diff
@@ -1210,20 +1210,20 @@ int DB::Object::get_state(const DoutPrefixProvider *dpp, RGWObjState **pstate, b
...
-  if (r < 0) {
-    return r;
-  }
-
-  *pmanifest = &(*astate->manifest);
-
-  return 0;
-}
-
+//int DB::Object::get_manifest(const DoutPrefixProvider *dpp, RGWObjManifest **pmanifest)
```
Comment (soumyakoduri) [inline src/rgw/store/dbstore/common/dbstore.cc:1213]: This function may be  removed completely from DBStore.

**Review by soumyakoduri — COMMENTED** [review]
Overall changes look good to me.

File: src/rgw/rgw_putobj_processor.cc [hunk 3]
```diff
@@ -304,13 +304,14 @@ int AtomicObjectProcessor::complete(size_t accounted_size,
...
+  head_obj->set_atomic();
 
   RGWRados::Object op_target(store->getRados(),
-		  head_obj->get_bucket()->get_info(),
-		  obj_ctx, head_obj->get_obj());
+		  head_obj->get_bucket(),
+		  obj_ctx, head_obj.get());
   RGWRados::Object::Write obj_op(&op_target);
 
+  ldpp_dout(dpp, 0) << "Luke 2 if_match=" << if_match << dendl;
```
[the remark this answers is not in the document]
Comment (dang) [reply src/rgw/rgw_putobj_processor.cc:314]: Thanks, missed these

File: src/rgw/store/dbstore/common/dbstore.h [hunk 4]
```diff
@@ -1800,7 +1798,7 @@ class DB {
           const rgw_obj& olh_obj, rgw_obj *target);
 
       int get_state(const DoutPrefixProvider *dpp, RGWObjState **pstate, bool follow_olh);
-      int get_manifest(const DoutPrefixProvider *dpp, RGWObjManifest **pmanifest);
+      /*int get_manifest(const DoutPrefixProvider *dpp, RGWObjManifest **pmanifest);*/
```
Comment (soumyakoduri) [inline src/rgw/store/dbstore/common/dbstore.h:1801]: nit: cleanup

File: src/rgw/rgw_sal.h [hunk 5]
```diff
@@ -980,37 +1011,62 @@ class Object {
     /** Set the ACL for this object */
     virtual int set_acl(const RGWAccessControlPolicy& acl) = 0;
     /** Mark further operations on this object as being atomic */
-    virtual void set_atomic(RGWObjectCtx* rctx) const = 0;
+    virtual void set_atomic() {
+      state.is_atomic = true;
```
[the remark this answers is not in the document]
Comment (dang) [reply src/rgw/rgw_sal.h:1015]: It seems to cause atomic RADOS ops to be used instead of non-atomic ops.

[the remark this answers is not in the document]
Comment (soumyakoduri) [reply src/rgw/rgw_sal.h:1015]: okay..thanks! I am unsure if its needed for other stores. But I guess that can be confirmed and cleaned up once the other stores are fully ready.

**Review by soumyakoduri — APPROVED** [review]
sal* changes look good to me.

**Comment by dang** [comment]
jenkins test make check

**Comment by dang** [comment]
jenkins test make check arm64

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
@andriytk can you review the MOTR bits?

**Review by andriytk — APPROVED** [review]
@dang all looks fine, thanks!

**Comment by dang** [comment]
Rebase still passes: 
https://pulpito.ceph.com/dang-2022-03-08_08:10:53-rgw-wip-dang-zipper-objctx-distro-default-smithi/

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test api

**Comment by dang** [comment]
jenkins test windows

**Comment by cdeshmukh** [comment]
Hi @dang, can you suggest anything, i believe you must of thought of same when doing these changes. We([cortx-rgw](https://github.com/Seagate/cortx-rgw)) are based out of quincy release right now and we will face issues when we rebase with main in future.

**Comment by cbodley** [comment]
> in the latest motr sal code we are using RGWObjectCtx to set the error msg from MotrObject::get_obj_state()

@andriytk @cdeshmukh this feels like a layering violation to me. `MotrObject::get_obj_state()` should't assume that a `req_state` even exists - these sal apis can be called through background threads like GC/LC, radosgw-admin commands, or test cases where there is no request context

if there's a need to pass more-specific errors back up to the calling `RGWOp`, we might look at adding new `ERR_` codes to rgw_common.h to represent those. but in this case, it looks like some RGWOps would just need to map `ENOENT`-> "The specified key does not exist." and `ERR_PRECONDITION_FAILED` -> "At least one of the pre-conditions you specified did not hold."

if i understand correctly, all radosgw backends would lack these error messages, so this shouldn't be solved in the motr layer

that linked PR mentions "copy object operation" specifically, so i'd start with `RGWCopyObj` and `RGWPutObj`. if we only want these messages in S3 and not Swift responses, they would go in `RGWCopyObj_ObjStore_S3::send_partial_response()` and `RGWPutObj_ObjStore_S3::send_response()`

**Comment by cdeshmukh** [comment]
> > in the latest motr sal code we are using RGWObjectCtx to set the error msg from MotrObject::get_obj_state()
> 
> @andriytk @cdeshmukh this feels like a layering violation to me. `MotrObject::get_obj_state()` should't assume that a `req_state` even exists - these sal apis can be called through background threads like GC/LC, radosgw-admin commands, or test cases where there is no request context

thanks @cbodley for the details, we raised this to understand proper ways of handling it.
And probably we can add additional check that `req_state` exists and then only proceed to use it, until we correct this approach & finish testing thoroughly. 

> 
> if there's a need to pass more-specific errors back up to the calling `RGWOp`, we might look at adding new `ERR_` codes to rgw_common.h to represent those. but in this case, it looks like some RGWOps would just need to map `ENOENT`-> "The specified key does not exist." and `ERR_PRECONDITION_FAILED` -> "At least one of the pre-conditions you specified did not hold."
> 
Yes, its needed for AWS S3 compliance

> if i understand correctly, all radosgw backends would lack these error messages, so this shouldn't be solved in the motr layer
>
Agree, our approach was for time being until we understand proper ways, for which this query was raised.

> that linked PR mentions "copy object operation" specifically, so i'd start with `RGWCopyObj` and `RGWPutObj`. if we only want these messages in S3 and not Swift responses, they would go in `RGWCopyObj_ObjStore_S3::send_partial_response()` and `RGWPutObj_ObjStore_S3::send_response()`

Yes, targeting in S3 alone for AWS compliance.

**Comment by cbodley** [comment]
> Yes, its needed for AWS S3 compliance

we've never really considered these error *messages* to be part of the s3 api for compliance purposes - only the error *codes* like `NoSuchKey`/`PreconditionFailed` along with the http status

these error messages are helpful for clients trying to debug issues so i definitely support adding more, but we shouldn't require them to match AWS' error messages verbatim

**Pull request merged**